parsers/src/main/java/vct/parsers/rewrite/KernelBodyRewriter.java

```diff
@@ -91,18 +91,22 @@ public void visit(Method m) {
             decls.add(rewrite(d));
         }
         Contract c = m.getContract();
+        if(c == null) {
+            c = new ContractBuilder().getContract(false);
+        }
         rewrite(c, icb);
+        icb.clearKernelInvariant();
```

Comment:
No, in fact C is the only place where we want a kernel_invariant :). CUDA kernels are just C methods with a `__global__` attribute (perhaps others). The code here is a little messy because it simultaneously generates the outer kernel contract, the invariant, the group parallel block contract and the kernel parallel block contract.